<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Andy Beck" />

<meta name="date" content="2024-12-06" />

<title>statistic_calibration</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">LSCI_pipeline</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">statistic_calibration</h1>
<h4 class="author">Andy Beck</h4>
<h4 class="date">2024-12-06</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-01-06
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>LSCI_pipeline/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20240916code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20240916)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20240916code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20240916)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomtheandybLSCIpipelinetreef74e31a5f394c9287ea2e3ffb63ce9915269d6a9targetblankf74e31aa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/theandyb/LSCI_pipeline/tree/f74e31a5f394c9287ea2e3ffb63ce9915269d6a9" target="_blank">f74e31a</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomtheandybLSCIpipelinetreef74e31a5f394c9287ea2e3ffb63ce9915269d6a9targetblankf74e31aa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/theandyb/LSCI_pipeline/tree/f74e31a5f394c9287ea2e3ffb63ce9915269d6a9" target="_blank">f74e31a</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rhistory
    Ignored:    .Rproj.user/
    Ignored:    data/1kgp/
    Ignored:    data/hgdp/
    Ignored:    data/reference/
    Ignored:    output/controls/
    Ignored:    output/figs/
    Ignored:    output/single_pos/
    Ignored:    output/singletons/
    Ignored:    output/slurm/
    Ignored:    output/two_pos/
    Ignored:    scratch/

Unstaged changes:
    Modified:   code/single_position_models_sc.py

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/statistic_calibration.Rmd</code>)
and HTML (<code>docs/statistic_calibration.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/theandyb/LSCI_pipeline/blob/f74e31a5f394c9287ea2e3ffb63ce9915269d6a9/analysis/statistic_calibration.Rmd" target="_blank">f74e31a</a>
</td>
<td>
Andy Beck
</td>
<td>
2025-01-06
</td>
<td>
wflow_publish(c("analysis/index.Rmd",
"analysis/statistic_calibration.Rmd",
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/theandyb/LSCI_pipeline/blob/bbc121da604d6d4a1c15c7b40e2a07cd83e16fe3/analysis/statistic_calibration.Rmd" target="_blank">bbc121d</a>
</td>
<td>
Andy Beck
</td>
<td>
2025-01-06
</td>
<td>
Add single chromosome code
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/theandyb/LSCI_pipeline/bbc121da604d6d4a1c15c7b40e2a07cd83e16fe3/docs/statistic_calibration.html" target="_blank">bbc121d</a>
</td>
<td>
Andy Beck
</td>
<td>
2025-01-06
</td>
<td>
Add single chromosome code
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<pre class="r"><code>library(tidyverse)</code></pre>
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.1     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>In this document, we will use simulation to evaluate the calibration
of our relative entropy statistic for the influence of local sequence
context on mutation rates.</p>
<div id="parameters" class="section level3">
<h3>Parameters</h3>
<p>The main parameter for each mutation subtype is the number of
singletons for that subtype. Here we’ll collect these counts for each
subtype across the five 1kGP super populations.</p>
<pre class="r"><code>df_singleton &lt;- data.frame(pop = rep(c(&quot;AFR&quot;, &quot;AMR&quot;, &quot;EAS&quot;, &quot;EUR&quot;, &quot;SAS&quot;), each = 9),
                           subtype = rep(c(&quot;AT_CG&quot;, &quot;AT_GC&quot;, &quot;AT_TA&quot;, &quot;GC_AT&quot;, &quot;GC_TA&quot;, &quot;GC_CG&quot;, &quot;cpg_GC_AT&quot;, &quot;cpg_GC_TA&quot;, &quot;cpg_GC_CG&quot;), times = 5),
                           n_s = c(1262888, 4178952, 1138869, 3821143, 1330996, 1452209, 1921101, 124142, 162111,
                                   537361, 1834264, 495037, 1681832, 563594, 623455,  827258,  55814, 67466,
                                   924484, 3190894, 888364, 2849828, 981610, 1084701, 1451677, 94483, 121589,
                                   710178, 2336815, 650119, 2124485, 731704, 821459,  1064404, 70878, 88486,
                                   897754, 2861359, 791460, 2686394, 890932, 1000478, 1346096, 89577, 113843))

df_singleton &lt;- df_singleton %&gt;%
  bind_rows({ df_singleton %&gt;%
    group_by(subtype) %&gt;%
    summarize(n_s = sum(n_s)) %&gt;%
    ungroup() %&gt;%
    mutate(pop = &quot;ALL&quot;) %&gt;%
    select(pop, subtype, n_s)})</code></pre>
<div id="nucleotide-frequencies-in-chr22" class="section level4">
<h4>Nucleotide frequencies in chr22</h4>
<p>For some of our simulations, we will want proportions of nucleotides.
While we could fiddle around with this, for a starting point let’s get
the frequency of each nucleotide in chromosome 22:</p>
<pre class="r"><code>df_22 &lt;- read_tsv(&quot;data/reference/gc100_22.bed&quot;,
                  col_names = c(&quot;chr&quot;, &quot;start&quot;, &quot;end&quot;, &quot;pct_at&quot;, &quot;pct_gc&quot;, &quot;n_A&quot;, &quot;n_C&quot;, &quot;n_G&quot;, &quot;n_T&quot;, &quot;n_N&quot;, &quot;n_O&quot;, &quot;len&quot;)) %&gt;%
  filter(n_N &lt; 50)</code></pre>
<pre><code>Rows: 508185 Columns: 12
── Column specification ────────────────────────────────────────────────────────
Delimiter: &quot;\t&quot;
chr  (1): chr
dbl (11): start, end, pct_at, pct_gc, n_A, n_C, n_G, n_T, n_N, n_O, len

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<pre class="r"><code>n_nuc_22 &lt;- df_22 %&gt;%
  ungroup() %&gt;%
  summarize(n_A = sum(n_A),
            n_C = sum(n_C),
            n_G = sum(n_G),
            n_T = sum(n_T)
            ) %&gt;%
  as.numeric()

prop_nuc_22 &lt;- n_nuc_22 / sum(n_nuc_22)</code></pre>
</div>
</div>
</div>
<div id="single-position-models" class="section level2">
<h2>Single Position Models</h2>
<p>In our first analysis, we will sample observed tables of singletons
and matched controls stratified by the nucleotide at a single flanking
position. We fill the following table:</p>
<table>
<thead>
<tr class="header">
<th>Nucleotide</th>
<th>Singletons</th>
<th>Controls</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>A</td>
<td><span class="math inline">\(n_{A,s}\)</span></td>
<td><span class="math inline">\(n_{A,c}\)</span></td>
</tr>
<tr class="even">
<td>C</td>
<td><span class="math inline">\(n_{C,s}\)</span></td>
<td><span class="math inline">\(n_{C,c}\)</span></td>
</tr>
<tr class="odd">
<td>G</td>
<td><span class="math inline">\(n_{G,s}\)</span></td>
<td><span class="math inline">\(n_{G,c}\)</span></td>
</tr>
<tr class="even">
<td>T</td>
<td><span class="math inline">\(n_{T,s}\)</span></td>
<td><span class="math inline">\(n_{T,c}\)</span></td>
</tr>
</tbody>
</table>
<p>using the following model:</p>
<p><span class="math display">\[
\log(n_{i,s}) = \lambda_0 + \sum_{j \in \{C,G,T\}} \lambda_j 1(i==j) +
\lambda_{\textrm{singleton}} 1(s==\textrm{singleton})
\]</span></p>
<p>We’ll look at both the deviance statistic (which has a theoretical
<span class="math inline">\(\chi^2\)</span> distribution with 3 degrees
of freedom) and the relative entropy statistic.</p>
<pre class="r"><code>simulate_1_pos_null &lt;- function(n_s, nuc_prop, c_s_ratio = 5){
  singletons &lt;- rmultinom(1, n_s, nuc_prop)
  controls &lt;- rmultinom(1, n_s * c_s_ratio, nuc_prop)
  
  test_df &lt;- data.frame(nuc = c(&quot;A&quot;, &quot;C&quot;, &quot;G&quot;, &quot;T&quot;),
                        singletons = singletons,
                        controls = controls) %&gt;%
    pivot_longer(singletons:controls, names_to = &quot;status&quot;, values_to = &quot;n&quot;)
  
  mod_obj &lt;- glm(n ~ nuc + status, test_df, family = poisson())
  
  dev_stat &lt;- deviance(mod_obj)
  re_val &lt;- dev_stat / (2 * n_s * (1 + c_s_ratio))
  return(data.frame(deviance = dev_stat, re = re_val))
}</code></pre>
<pre class="r"><code>n_s &lt;- df_singleton %&gt;%
  filter(pop == &quot;ALL&quot;, subtype == &quot;AT_GC&quot;) %&gt;%
  pull(n_s)

n_sim &lt;- 2000

set.seed(1848)
df_sim &lt;- vector(mode = &quot;list&quot;, length = n_sim)
for(i in 1:n_sim){
  df_sim[[i]] &lt;- simulate_1_pos_null(n_s, prop_nuc_22)
}

df_sim &lt;- df_sim %&gt;% bind_rows()</code></pre>
<p>Out of our 2000 simulations, how many false positive statistics do we
observe?</p>
<pre class="r"><code>sum(df_sim$deviance &gt; qchisq(0.95, 3))</code></pre>
<pre><code>[1] 90</code></pre>
<p>This is incredibly close to 0.05. Now, among the statistics which
were significant, what does their distribution look like?</p>
<pre class="r"><code>df_sim %&gt;%
  filter(deviance &gt; qchisq(0.95, 3)) %&gt;%
  pull(deviance) %&gt;%
  {plot(density(.))}</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>How does the overall distribution of statistics look relative to the
theoretical <span
class="math inline">\(\chi^2_{\textrm{df}=3}\)</span>?</p>
<pre class="r"><code>ggplot() +
  aes(sample = df_sim$deviance) +
  stat_qq(distribution = qchisq, dparams = 3) +
  stat_qq_line(col = &quot;blue&quot;, distribution = qchisq, dparams = 3) +
  xlab(&quot;Quantile&quot;) + ylab(&quot;Deviance&quot;)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>And finally, what does our density look like?</p>
<pre class="r"><code>df_sim %&gt;%
  ggplot(aes(x = deviance)) +
  geom_density() +
  stat_function(fun = dchisq, args = c(3), n = 1000, colour = &quot;red&quot;) + 
  geom_vline(xintercept = 1, colour = &quot;red&quot;, linetype = 2)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>MASS::fitdistr(df_sim$deviance, &quot;chi-squared&quot;, start = list(&quot;df&quot; = 4))</code></pre>
<pre><code>Warning in stats::optim(x = c(1.08542398408702, 1.78680277324611, 4.7001163318746, : one-dimensional optimization by Nelder-Mead is unreliable:
use &quot;Brent&quot; or optimize() directly</code></pre>
<pre><code>       df    
  2.98281250 
 (0.04607813)</code></pre>
</div>
<div id="alternative-simulation-strategy" class="section level2">
<h2>Alternative simulation strategy</h2>
<p>Here we’ll simulate directly from the poisson distribution. We’ll
place normal priors on the nucleotide parameters, and fix the status
parameter at <span class="math inline">\(log(\frac{1}{5})\)</span></p>
<pre class="r"><code>simulate_1_pos_null_poisson &lt;- function(n_s, nuc_prop, c_s_ratio = 5){
  
  beta_s = log(1/c_s_ratio)
  beta_0 = log(n_s * c_s_ratio * nuc_prop[1])
  beta_c = log(nuc_prop[2] / nuc_prop[1])
  beta_g = log(nuc_prop[3] / nuc_prop[1])
  beta_t = log(nuc_prop[4] / nuc_prop[1])
  
  # theta_0 &lt;- rnorm(1, beta_0, 1/10)
  # theta_c &lt;- rnorm(1, beta_c, 1/(10 ^ (ceiling(-1*log10(-1 * beta_c)) + 2) ))
  # theta_g &lt;- rnorm(1, beta_g, 1/(10 ^ (ceiling(-1*log10(-1 * beta_g)) + 2) ))
  # theta_t &lt;- rnorm(1, beta_t, 1/(10 ^ (ceiling(-1*log10(-1 * beta_t)) + 2) ))
  
  controls &lt;- c(rpois(1, exp(beta_0)), rpois(1, exp(beta_0 + beta_c)), rpois(1, exp(beta_0+ beta_g)),  rpois(1, exp(beta_0+ beta_t)) )
  #controls &lt;- c(controls, n_s*c_s_ratio - sum(controls))
  controls[controls&lt;0] &lt;- 0
  
  singletons &lt;- c(rpois(1, exp(beta_0 + beta_s)), rpois(1, exp(beta_0 + beta_c + beta_s)), rpois(1, exp(beta_0+ beta_g + beta_s)), rpois(1, exp(beta_0+ beta_t + beta_s)))
  #singletons &lt;- c(singletons, n_s - sum(singletons))
  singletons[singletons&lt;0] &lt;- 0
  
  test_df &lt;- data.frame(nuc = c(&quot;A&quot;, &quot;C&quot;, &quot;G&quot;, &quot;T&quot;),
                        singletons = singletons,
                        controls = controls) %&gt;%
    pivot_longer(singletons:controls, names_to = &quot;status&quot;, values_to = &quot;n&quot;)
  
  mod_obj &lt;- glm(n ~ nuc + status, test_df, family = poisson())
  
  dev_stat &lt;- deviance(mod_obj)
  re_val &lt;- dev_stat / (2 * n_s * (1 + c_s_ratio))
  return(data.frame(deviance = dev_stat, re = re_val))
}</code></pre>
<pre class="r"><code>n_s &lt;- df_singleton %&gt;%
  filter(pop == &quot;ALL&quot;, subtype == &quot;AT_GC&quot;) %&gt;%
  pull(n_s)

n_sim &lt;- 2000

set.seed(1848)
df_sim2 &lt;- vector(mode = &quot;list&quot;, length = n_sim)
for(i in 1:n_sim){
  df_sim2[[i]] &lt;- simulate_1_pos_null_poisson(n_s, prop_nuc_22)
}</code></pre>
<pre><code>Warning: glm.fit: algorithm did not converge</code></pre>
<pre class="r"><code>df_sim2 &lt;- df_sim2 %&gt;% bind_rows()</code></pre>
<pre class="r"><code>sum(df_sim2$deviance &gt; qchisq(0.95, 3))</code></pre>
<pre><code>[1] 98</code></pre>
<pre class="r"><code>df_sim2 %&gt;%
  ggplot(aes(x = deviance)) +
  geom_density() +
  stat_function(fun = dchisq, args = c(3), n = 1000, colour = &quot;red&quot;) + 
  geom_vline(xintercept = 1, colour = &quot;red&quot;, linetype = 2)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-14-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggplot() +
  aes(sample = df_sim2$deviance) +
  stat_qq(distribution = qchisq, dparams = 3) +
  stat_qq_line(col = &quot;blue&quot;, distribution = qchisq, dparams = 3) +
  xlab(&quot;Quantile&quot;) + ylab(&quot;Deviance&quot;)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-15-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>MASS::fitdistr(df_sim2$deviance, &quot;chi-squared&quot;, start = list(&quot;df&quot; = 3))</code></pre>
<pre><code>Warning in stats::optim(x = c(3.06721923759667, 2.34851958708578, 3.1292682101448, : one-dimensional optimization by Nelder-Mead is unreliable:
use &quot;Brent&quot; or optimize() directly</code></pre>
<pre><code>       df    
  2.93847656 
 (0.04562023)</code></pre>
</div>
<div id="poisson-simulation-with-fixed-margins" class="section level2">
<h2>Poisson Simulation with Fixed Margins</h2>
<pre class="r"><code>simulate_1_pos_null_fixM &lt;- function(n_s, nuc_prop, c_s_ratio = 5){
  
  beta_s = log(1/c_s_ratio)
  beta_0 = log(n_s * c_s_ratio * nuc_prop[1])
  beta_c = log(nuc_prop[2] / nuc_prop[1])
  beta_g = log(nuc_prop[3] / nuc_prop[1])
  beta_t = log(nuc_prop[4] / nuc_prop[1])
  
  controls &lt;- c(rpois(1, exp(beta_0)), rpois(1, exp(beta_0 + beta_c)), rpois(1, exp(beta_0+ beta_g)))
  controls &lt;- c(controls, n_s*c_s_ratio - sum(controls))
  #controls[controls&lt;0] &lt;- 0
  
  singletons &lt;- c(rpois(1, exp(beta_0 + beta_s)), rpois(1, exp(beta_0 + beta_c + beta_s)), rpois(1, exp(beta_0+ beta_g + beta_s)))
  singletons &lt;- c(singletons, n_s - sum(singletons))
  #singletons[singletons&lt;0] &lt;- 0
  
  test_df &lt;- data.frame(nuc = c(&quot;A&quot;, &quot;C&quot;, &quot;G&quot;, &quot;T&quot;),
                        singletons = singletons,
                        controls = controls) %&gt;%
    pivot_longer(singletons:controls, names_to = &quot;status&quot;, values_to = &quot;n&quot;)
  
  mod_obj &lt;- glm(n ~ nuc + status, test_df, family = poisson())
  
  dev_stat &lt;- deviance(mod_obj)
  re_val &lt;- dev_stat / (2 * n_s * (1 + c_s_ratio))
  return(data.frame(deviance = dev_stat, re = re_val))
}

simulate_1_pos_null_fixM_altM &lt;- function(n_s, nuc_prop, c_s_ratio = 5){
  
  beta_s = log(1/c_s_ratio)
  beta_0 = log(n_s * c_s_ratio * nuc_prop[1])
  beta_c = log(nuc_prop[2] / nuc_prop[1])
  beta_g = log(nuc_prop[3] / nuc_prop[1])
  beta_t = log(nuc_prop[4] / nuc_prop[1])
  
  controls &lt;- c(rpois(1, exp(beta_0)), rpois(1, exp(beta_0 + beta_c)), rpois(1, exp(beta_0+ beta_g)))
  controls &lt;- c(controls, n_s*c_s_ratio - sum(controls))
  #controls[controls&lt;0] &lt;- 0
  
  singletons &lt;- c(rpois(1, exp(beta_0 + beta_s)), rpois(1, exp(beta_0 + beta_c + beta_s)), rpois(1, exp(beta_0+ beta_g + beta_s)))
  singletons &lt;- c(singletons, n_s - sum(singletons))
  #singletons[singletons&lt;0] &lt;- 0
  
  test_df &lt;- data.frame(nuc = c(&quot;A&quot;, &quot;C&quot;, &quot;G&quot;, &quot;T&quot;),
                        singletons = singletons,
                        controls = controls) %&gt;%
    pivot_longer(singletons:controls, names_to = &quot;status&quot;, values_to = &quot;n&quot;) %&gt;%
    left_join(data.frame(status = c(&quot;singletons&quot;, &quot;controls&quot;), col_sum = c(n_s, n_s*5)), by = &quot;status&quot;)
  
  mod_obj &lt;- glm(n ~ nuc + offset(log(col_sum)), test_df, family = poisson())
  
  dev_stat &lt;- deviance(mod_obj)
  re_val &lt;- dev_stat / (2 * n_s * (1 + c_s_ratio))
  return(data.frame(deviance = dev_stat, re = re_val))
}</code></pre>
<pre class="r"><code>n_s &lt;- df_singleton %&gt;%
  filter(pop == &quot;ALL&quot;, subtype == &quot;AT_GC&quot;) %&gt;%
  pull(n_s)

n_sim &lt;- 2000

set.seed(1848)
df_sim3 &lt;- vector(mode = &quot;list&quot;, length = n_sim)
for(i in 1:n_sim){
  df_sim3[[i]] &lt;- simulate_1_pos_null_fixM(n_s, prop_nuc_22)
}

df_sim3 &lt;- df_sim3 %&gt;% bind_rows()

set.seed(1848)
df_sim4 &lt;- vector(mode = &quot;list&quot;, length = n_sim)
for(i in 1:n_sim){
  df_sim4[[i]] &lt;- simulate_1_pos_null_fixM_altM(n_s, prop_nuc_22)
}

df_sim4 &lt;- df_sim4 %&gt;% bind_rows()</code></pre>
<pre class="r"><code>sum(df_sim3$deviance &gt; qchisq(0.95, 3))</code></pre>
<pre><code>[1] 489</code></pre>
<pre class="r"><code>sum(df_sim4$deviance &gt; qchisq(0.95, 4))</code></pre>
<pre><code>[1] 370</code></pre>
<pre class="r"><code>df_sim3 %&gt;%
  ggplot(aes(x = deviance)) +
  geom_density() +
  stat_function(fun = dchisq, args = c(3), n = 1000, colour = &quot;red&quot;) +
  geom_vline(xintercept = 1, colour = &quot;red&quot;, linetype = 2)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-20-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code># 
df_sim4 %&gt;%
  ggplot(aes(x = deviance)) +
  geom_density() +
  stat_function(fun = dchisq, args = list(&quot;df&quot; = 4, &quot;ncp&quot; = 0), n = 1000, colour = &quot;red&quot;) +
  geom_vline(xintercept = 2, colour = &quot;red&quot;, linetype = 2)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-20-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggplot() +
  aes(sample = df_sim3$deviance) +
  stat_qq(distribution = qchisq, dparams = 3) +
  stat_qq_line(col = &quot;blue&quot;, distribution = qchisq, dparams = list(&quot;df&quot; = 4, &quot;ncp&quot; = 0)) +
  xlab(&quot;Quantile&quot;) + ylab(&quot;Deviance&quot;)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>ggplot() +
  aes(sample = df_sim3$deviance) +
  stat_qq(distribution = qchisq, dparams = 4) +
  stat_qq_line(col = &quot;blue&quot;, distribution = qchisq, dparams = 4) +
  xlab(&quot;Quantile&quot;) + ylab(&quot;Deviance&quot;)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-21-2.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>opt_1 &lt;- MASS::fitdistr(df_sim3$deviance, &quot;chi-squared&quot;, start = list(&quot;df&quot; = 4, &quot;ncp&quot; = 3))</code></pre>
<pre><code>Warning in densfun(x, parm[1], parm[2], ...): NaNs produced
Warning in densfun(x, parm[1], parm[2], ...): NaNs produced
Warning in densfun(x, parm[1], parm[2], ...): NaNs produced
Warning in densfun(x, parm[1], parm[2], ...): NaNs produced
Warning in densfun(x, parm[1], parm[2], ...): NaNs produced</code></pre>
<pre class="r"><code>opt_1$estimate</code></pre>
<pre><code>      df      ncp 
1.827676 3.764550 </code></pre>
<pre class="r"><code>opt_2 &lt;- MASS::fitdistr(df_sim3$deviance, &quot;chi-squared&quot;, start = list(&quot;ncp&quot; = 1), df=4)</code></pre>
<pre><code>Warning in stats::optim(x = c(3.37229132652169, 7.1735898455986, 4.08473925053249, : one-dimensional optimization by Nelder-Mead is unreliable:
use &quot;Brent&quot; or optimize() directly</code></pre>
<pre class="r"><code>opt_2$estimate</code></pre>
<pre><code>     ncp 
1.526563 </code></pre>
<p>Below we plot the density of the deviances along with the non-central
<span class="math inline">\(\chi^2\)</span> distribution with both df
and ncp parameters optimized:</p>
<pre class="r"><code>df_sim3 %&gt;%
  ggplot(aes(x = deviance)) +
  geom_density() +
  stat_function(fun = dchisq, args = list(&quot;df&quot; = opt_1$estimate[&quot;df&quot;], &quot;ncp&quot; = opt_1$estimate[&quot;ncp&quot;]), n = 1000, colour = &quot;red&quot;)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-24-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Now plot the estimated distribution where we’ve fixed the df to be
4:</p>
<pre class="r"><code>df_sim3 %&gt;%
  ggplot(aes(x = deviance)) +
  geom_density() +
  stat_function(fun = dchisq, args = list(&quot;df&quot; = 4, &quot;ncp&quot; = opt_2$estimate[&quot;ncp&quot;]), n = 1000, colour = &quot;red&quot;)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-25-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>And finally the distribution where we use a quasi-method-of-moments
estimator of the ncp assuming df = 3</p>
<pre class="r"><code>ncp_est &lt;- mean(df_sim3$deviance) - 4
df_sim3 %&gt;%
  ggplot(aes(x = deviance)) +
  geom_density() +
  stat_function(fun = dchisq, args = list(&quot;df&quot; = 4, &quot;ncp&quot; = ncp_est), n = 1000, colour = &quot;red&quot;)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-26-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>Lastly, to satisfy my curiosity, if we fix df = 4 and find ncp based
on variance formula for the nc-chisq:</p>
<pre class="r"><code>ncp_var_est &lt;- (var(df_sim3$deviance) - 8) / 4
df_sim3 %&gt;%
  ggplot(aes(x = deviance)) +
  geom_density() +
  stat_function(fun = dchisq, args = list(&quot;df&quot; = 4, &quot;ncp&quot; = ncp_var_est), n = 1000, colour = &quot;red&quot;)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-27-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>df_sim3 %&gt;%
  ggplot(aes(x = deviance)) +
  geom_density() +
  stat_function(fun = dchisq, args = c(3, ncp_est), n = 1000, colour = &quot;red&quot;)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-28-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>df_sim3 %&gt;%
  ggplot(aes(x = deviance)) +
  geom_density() +
  stat_function(fun = dchisq, args = c(3, ncp_est), n = 1000, colour = &quot;red&quot;)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-29-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="fit-poisson-model-with-offset" class="section level2">
<h2>Fit poisson model with offset</h2>
<p>Here I want to see what happens with our single position test if we
use an offset to condition on fix totals for singleton/control
categories:</p>
<pre class="r"><code>singletons &lt;- rmultinom(1, n_s, prop_nuc_22) #n_s * prop_nuc_22
controls &lt;- rmultinom(1, n_s*5, prop_nuc_22 + c(-0.1, 0.1, -0.02, 0.02))

test_df &lt;- data.frame(nuc = c(&quot;A&quot;, &quot;C&quot;, &quot;G&quot;, &quot;T&quot;),
                      singletons = singletons,
                      controls = controls) %&gt;%
  pivot_longer(singletons:controls, names_to = &quot;status&quot;, values_to = &quot;n&quot;)

test_df &lt;- test_df %&gt;%
  left_join(data.frame(status = c(&quot;singletons&quot;, &quot;controls&quot;), col_sum = c(n_s, n_s*5)), by = &quot;status&quot;)

mod_obj &lt;- glm(n ~ nuc + status, test_df, family = poisson)
mod_obj2 &lt;- glm(n ~ nuc + offset(log(col_sum)), test_df, family = poisson)
mod_obj3 &lt;- glm(n ~ nuc * status + offset(log(col_sum)), test_df, family=poisson)</code></pre>
<div id="impact-of-sample-size-on-distribution" class="section level3">
<h3>Impact of sample size on distribution</h3>
<p>Let’s look at the distribution of the statistics in a simulation
using a smaller sample size:</p>
<pre class="r"><code>n_s &lt;- df_singleton %&gt;%
  pull(n_s) %&gt;%
  min()

n_sim &lt;- 2000

set.seed(1848)
df_sim &lt;- vector(mode = &quot;list&quot;, length = n_sim)
for(i in 1:n_sim){
  df_sim[[i]] &lt;- simulate_1_pos_null_fixM(n_s, prop_nuc_22)
}

df_sim &lt;- bind_rows(df_sim)</code></pre>
<pre class="r"><code>sum(df_sim$deviance &gt; qchisq(0.95, 3))</code></pre>
<pre><code>[1] 473</code></pre>
<pre class="r"><code>ggplot() +
  aes(sample = df_sim$deviance) +
  stat_qq(distribution = qchisq, dparams = 4) +
  stat_qq_line(col = &quot;blue&quot;, distribution = qchisq, dparams = 3) +
  xlab(&quot;Quantile&quot;) + ylab(&quot;Deviance&quot;)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-33-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>And finally, what does our density look like?</p>
<pre class="r"><code>df_sim %&gt;%
  ggplot(aes(x = deviance)) +
  geom_density() +
  stat_function(fun = dchisq, args = list(&quot;df&quot; = 4), n = 1000, colour = &quot;red&quot;) + 
  geom_vline(xintercept = 2, colour = &quot;red&quot;, linetype = 2)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-34-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>df_sim %&gt;%
  mutate(type = &quot;small&quot;) %&gt;%
  bind_rows({df_sim3 %&gt;% mutate(type = &quot;big&quot;)}) %&gt;%
  ggplot(aes(x = deviance, color = type)) +
  geom_density()</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-35-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>df_sim %&gt;%
  mutate(type = &quot;small&quot;) %&gt;%
  bind_rows({df_sim3 %&gt;% mutate(type = &quot;big&quot;)}) %&gt;%
  ggplot() + 
  aes(sample = deviance, color = type) + 
  stat_qq(distribution = qchisq, dparams = 4) +
  stat_qq_line(col = &quot;blue&quot;, distribution = qchisq, dparams = 3) +
  xlab(&quot;Quantile&quot;) + ylab(&quot;Deviance&quot;)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-35-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="varyinig-nucleotide-distributions" class="section level3">
<h3>Varyinig nucleotide distributions</h3>
<pre class="r"><code>nuc_prop_high_gh &lt;- c(0.15,0.34,0.36,0.15)

n_s &lt;- df_singleton %&gt;%
  pull(n_s) %&gt;%
  min()

n_sim &lt;- 2000

set.seed(1848)
df_sim_gc &lt;- vector(mode = &quot;list&quot;, length = n_sim)
for(i in 1:n_sim){
  df_sim_gc[[i]] &lt;- simulate_1_pos_null_fixM(n_s, nuc_prop_high_gh)
}

df_sim_gc &lt;- df_sim_gc %&gt;% bind_rows()</code></pre>
<pre class="r"><code>sum(df_sim_gc$deviance &gt; qchisq(0.95, 4))</code></pre>
<pre><code>[1] 608</code></pre>
<pre class="r"><code>ggplot() +
  aes(sample = df_sim_gc$deviance) +
  stat_qq(distribution = qchisq, dparams = 4) +
  stat_qq_line(col = &quot;blue&quot;, distribution = qchisq, dparams = 4) +
  xlab(&quot;Quantile&quot;) + ylab(&quot;Deviance&quot;)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-38-1.png" width="672" style="display: block; margin: auto;" /></p>
<p>And finally, what does our density look like?</p>
<pre class="r"><code>df_sim_gc %&gt;%
  ggplot(aes(x = deviance)) +
  geom_density() +
  stat_function(fun = dchisq, args = list(&quot;df&quot; = 4), n = 1000, colour = &quot;red&quot;) + 
  geom_vline(xintercept = 2, colour = &quot;red&quot;, linetype = 2)</code></pre>
<p><img src="figure/statistic_calibration.Rmd/unnamed-chunk-39-1.png" width="672" style="display: block; margin: auto;" /></p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.4.2 (2024-10-31)
Platform: x86_64-pc-linux-gnu
Running under: Ubuntu 20.04.6 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3;  LAPACK version 3.9.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: America/New_York
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lubridate_1.9.3 forcats_1.0.0   stringr_1.5.1   dplyr_1.1.4    
 [5] purrr_1.0.2     readr_2.1.5     tidyr_1.3.1     tibble_3.2.1   
 [9] ggplot2_3.5.1   tidyverse_2.0.0 workflowr_1.7.1

loaded via a namespace (and not attached):
 [1] sass_0.4.9        utf8_1.2.4        generics_0.1.3    stringi_1.8.4    
 [5] hms_1.1.3         digest_0.6.37     magrittr_2.0.3    timechange_0.3.0 
 [9] evaluate_1.0.1    grid_4.4.2        fastmap_1.2.0     rprojroot_2.0.4  
[13] jsonlite_1.8.9    processx_3.8.4    whisker_0.4.1     ps_1.8.1         
[17] promises_1.3.0    httr_1.4.7        fansi_1.0.6       scales_1.3.0     
[21] jquerylib_0.1.4   cli_3.6.3         crayon_1.5.3      rlang_1.1.4      
[25] bit64_4.0.5       munsell_0.5.1     withr_3.0.2       cachem_1.1.0     
[29] yaml_2.3.10       parallel_4.4.2    tools_4.4.2       tzdb_0.4.0       
[33] colorspace_2.1-1  httpuv_1.6.15     vctrs_0.6.5       R6_2.5.1         
[37] lifecycle_1.0.4   git2r_0.33.0      bit_4.5.0.1       fs_1.6.5         
[41] MASS_7.3-61       vroom_1.6.5       archive_1.1.8     pkgconfig_2.0.3  
[45] callr_3.7.6       pillar_1.9.0      bslib_0.8.0       later_1.3.2      
[49] gtable_0.3.6      glue_1.8.0        Rcpp_1.0.13-1     xfun_0.49        
[53] tidyselect_1.2.1  rstudioapi_0.17.1 knitr_1.49        farver_2.1.2     
[57] htmltools_0.5.8.1 labeling_0.4.3    rmarkdown_2.29    compiler_4.4.2   
[61] getPass_0.2-4    </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
